0010-endpoint-metadata-extension/README.md

```diff
@@ -0,0 +1,687 @@
+  - [Get metadata](#get-metadata)
+    - [Get metadata request](#get-metadata-request)
+    - [Get metadata response](#get-metadata-response)
+  - [Get metadata keys](#get-metadata-keys)
+    - [Get metadata keys request](#get-metadata-keys-request)
+    - [Get metadata keys response](#get-metadata-keys-response)
+  - [Delete metadata key](#delete-metadata-key)
+    - [Delete metadata key request](#delete-metadata-key-request)
+    - [Delete metadata key response](#delete-metadata-key-response)
+
+
+## Introduction
+
+The Endpoint Metadata Extension (EPMDX) protocol is an endpoint-aware [Kaa Protocol](/0001-kaa-protocol/README.md) extension.
+
+It is intended to manage endpoint metadata.
+
+## Requirements and constraints
+
+### Metadata keys
+EPMDX protocol supports only string type of metadata keys.
+All UTF-8 characters are supported in metadata key names.
+Metadata key must contain at least one non-whitespace character.
+For example, `longitude`, `street`, `name`, `$address 2` are valid metadata keys.
+No duplicate metadata keys allowed.
+### Metadata values
+Valid metadata value type is any JSON type:
+- `boolean`
+- `string`
+- `number`
+- `null`
+- `object`
+- `array`
+
+Example endpoint metadata:
+
+```json
+{
+  "name":"Sensor 1",
+  "OS name":"Linux",
+  "OS version":"4.2.9",
+  "cores":2,
+  "ssd":true,
+  "location":{
+    "latitude":27.664827,
+    "longitude":-81.515754
+  },
+  "supportedFirmwareVersions":[
+    "2.0.0",
+    "2.0.1"
+  ]
+}
+```
+### Supported operations
+
+EPMDX protocol must support the following operations:
+- Get metadata key-value pairs (all or for a certain keys)
+- Get a list of metadata keys
+- Update metadata key-value pairs (all or partial update)
+- Delete a single metadata key value pair by key name
+
+## Design
+
+### Full metadata update
+
+Create or update endpoint metadata operation.
+In case of update the provided metadata will override all existing metadata key-value pairs.
+Use [partial update operation](#partial-metadata-update) to avoid removing an existing values.
+
+#### Full metadata update request
+
+Client MUST send requests with the following extension-specific [resource path](/0001-kaa-protocol/README.md#resource-path-format) part in order to receive the metadata key-value pairs:
+
+  `<endpoint_token>/update`
+
+  where `endpoint_token` identifies the endpoint.
+
+The request payload MUST be a JSON-encoded object with the following [JSON schema](http://json-schema.org/) ([file](./schemas/update-metadata-request.schema.json)):
+
+```json
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "description": "10/EPMDX update metadata request",
+  "type": "object",
+  "properties": {
+    "id": {
+      "type": "integer",
+      "description": "The message identifier used to match server response to the request"
+    },
+    "metadata": {
+      "type": "object",
+      "minProperties": 1,
+      "patternProperties": {
+        "\\S+": {
```

Comment:
According to JSON schema, you can use `{}` as a catch-all type. (Instead of listing all types.)